bgpd/bgp_evpn.c

```diff
@@ -1747,10 +1796,19 @@ static int update_evpn_route(struct bgp *bgp, struct bgpevpn *vpn,
 	assert(pi);
 	attr_new = pi->attr;
 
+	/* lock ri to prevent freeing in evpn_route_select_install */
+	bgp_path_info_lock(pi);
 	/* Perform route selection; this is just to set the flags correctly
 	 * as local route in the VNI always wins.
 	 */
 	evpn_route_select_install(bgp, vpn, rn);
+	/*
+	 * if the new local route was not selected evict it and tell zebra
+	 * to add the best remote dest
+	 */
+	evpn_cleanup_local_non_best_route(bgp, vpn, rn, pi, &route_change);
```

Comment:
I am not following this change.  evpn_route_select_install should have run the best path algorithm and selected the best path (remote route in this case) already. So why do we need to do this again? If not modifying the evpn_route_select_install to select the best route would be a correct way rather than having another function to fix some thing that was done by evpn_route_select_install.

Reply:
evpn_route_select_install picks a best path for a rn in a ri independent fashion i.e. it doesn't care if a ri is being added/modified/deleted. also it does't care if it is being run because of local updates or imported route updates. this fixup is very specific to addition of a local ri so it cannot be done within evpn_route_select_install (that api is already overloaded IMO). also evpn_route_select_install only updates zebra if the currently selected path changes; in this specific handling that didn't happen. i get your hesitation about a follow-fixup but think of this more as NAKing zebra's add and re-instating bgp's decision. The main rationale is that bgpd shouldn't have a local path in it's route table unless that path is the best path; neither should zebra.